pip/pip-355.md

```diff
@@ -0,0 +1,39 @@
+# PIP-355: Enhancing Broker-Level Metrics for Pulsar
+
+# Background Knowledge
+Pulsar provides broker-level, namespace-level, and topic-level metrics to monitor and analyze the behavior of the Pulsar service. These metrics are accessible through the Prometheus metrics endpoint. Detailed explanations of all metrics can be found on the Pulsar website: [Pulsar Metrics Reference](https://pulsar.apache.org/docs/3.2.x/reference-metrics/)
+
+# Motivation
+Within Pulsar's current metrics framework, the `pulsar_out_bytes_total` metric is utilized to expose the total bytes dispatched by the broker to consumers. However, there are notable limitations and challenges associated with this metric:
+- Inclusion of system subscriptions in the total bytes out, alongside user subscriptions, complicates accurate calculation of user-specific data.
+- The granularity of the metric (namespace-level vs. topic-subscription level) impacts the scalability and resource consumption when calculating cluster-level total out bytes.
+
+# Goals
+This proposal aims to address the following objectives:
+- Simplify the process of calculating cluster-level total out bytes.
+- Enable the calculation of total out bytes dispatched to system subscriptions.
+
+# High-Level Design
+To achieve the outlined goals, the proposal introduces three new broker-level metrics:
+- `pulsar_broker_out_bytes_total`: Represents the total out bytes dispatched by the broker to consumers, encompassing both user and system subscriptions.
```

Comment:
I'm trying to think how this would be written in OTel, to compare.
In OTel we won't have broker level metrics, since we will always have namespace level metrics - I believe @dragosvictor is working on adding them).
Assuming `pulsar.broker.namespace.outgoing.size`, then it may be reasonable to have another attribute named `pulsar.system.topic=true|false`.

In Prometheus, we can call the label `system_topic=true|false`.

The reason I was thinking about it is that it's unlikely to put anything after `_total`, and the second reason is that nobody will understand `_internal` without going to the docs.

Reply:
That sounds a good idea

For prometheus:
```
pulsar_broker_out_bytes_total {system_topic="true | false"}
pulsar_broker_in_bytes_total {system_topic="true | false"}
```

Is that correct? Or we can should use the same name, for example

```
pulsar_broker_out_size {system_topic="true | false"}
pulsar_broker_in_size {system_topic="true | false"}
```

Reply:
Seconded. If we can use attributes instead of multiple metric names, we can aid the discovery and understanding of these metrics, as well as offer an easy path for future extension if desired. Continuing with that second point, maybe instead of boolean label `system_topic`, we can have a string label `subscription_scope` with values `system` and `user`?

Reply:
Adding a label instead of a new metric name makes sense.

For `pulsar_broker_out_bytes_total`, adding the label `subscription_scope`, which is more reasonable because we may add system cursors on any topics and those outgoing traffic should be treated as system traffic

For `pulsar_broker_in_size`, adding the label `system_topic` is more reasonable.

Reply:
Maybe `topic_scope` instead of `subscription_scope`, such that we can have the same label across the two I/O directions.

Reply:
`topic_scope` is also a little confusing for `pulsar_broker_out_bytes_total`. In general understanding, `topic_scope=system` means only taking system topic dispatch traffic into account. However, it means any topics with system cursors dispatch traffic.

Reply:
I would like those labels we add to be consistent across in and out. Instead of using two new terms "system" and "scope", how about we add two boolean labels:
1. `system_topic` when this is a system (internal) topic.
2. `system_subscription` when this is a system (internal) subscription.

Should look like this:
```
pulsar_broker_in_bytes_total{cluster=, topic=, system_topic=true}
pulsar_broker_out_bytes_total{cluster=, topic=, subscription=, system_subscription=true}
pulsar_broker_out_bytes_total{cluster=, topic=, subscription=, system_subscription=true, system_topic=true}
```

WDYT?

Reply:
From the usage perspective, we only care about the `system_subscription` in `pulsar_broker_out_bytes_total`, and do not need system_topic, which will make the metric generation complex.
The following two metric is enough:
```
pulsar_broker_in_bytes_total{cluster=, topic=, system_topic=true}
pulsar_broker_out_bytes_total{cluster=, topic=, subscription=, system_subscription=true}
```
But it will fall back to this pair of metrics having different labels. I'm not sure if it is acceptable.

Reply:
@asafm My idea was to introduce an attribute that can be easily extended in the future if needed, which is why I proposed a string attribute. A boolean restricts this, and can introduce various combinations which might not make sense in practice (for instance, `system_subscription=true, system_topic=false` doesn't look probable IMO).

WDYT?

Reply:
@hangc0276 wrote that there exists a situation that a normal topic will have a system subscription (maybe replication has that use case?). In that case, the attribute `topic_scope` will be confusing and doesn't match what you wish to convey: This is a normal topic and the subscription is a system one. The boolean labels help to differentiate between those use cases.

Reply:
As discussed with @dragosvictor, we will use the label name `scope` instead of `topic_scope`, for example:
```
pulsar_broker_in_bytes_total{cluster=, topic=, scope="system | user"}
pulsar_broker_out_bytes_total{cluster=, topic=, scope="system | user"}
```
I agree with @asafm, it will be a little confusing because the meaning of `scope` in `pulsar_broker_in_bytes_total` and `pulsar_broker_out_bytes_total` is different. 
In `pulsar_broker_in_bytes_total`, it means system topic or user normal topic
In `pulsar_broker_out_bytes_total`, it means system cursor or user cursor.

In order to make the meaning of label in sync and easy to understand, I perfer @asafm's idea to use two separate labels. Just like:
```
pulsar_broker_in_bytes_total{cluster=, system_topic="true | false" }
pulsar_broker_out_bytes_total{cluster=, system_subscription=" true | false"}
```

Reply:
@dragosvictor Do you have any concerns?

Reply:
Sounds good to me :+1: